auth-server/src/services/auth.ts

```diff
@@ -486,16 +624,78 @@ export default class AuthService {
 			.patch({ valid: false })
 			.where({ user_id: userId });
 
-		const verifyToken = await EmailVerificationToken
+		await this.sendVerificationTokenMail(user);
+	}
+
+	public async SetCredentialsStart (address: string): Promise<string> {
+		const addressObj = await Address
 			.query()
-			.allowInsert('[token, user_id, valid]')
-			.insert({
-				token: uuid(),
-				user_id: user.id,
-				valid: true
-			});
+			.where('address', address)
+			.first();
+
+		if (!addressObj) {
+			throw new ForbiddenError(messages.ADDRESS_NOT_FOUND);
+		}
+
+		const signMessage = uuid();
+
+		await redisSetex(getSetCredentialsKey(address), ADDRESS_LOGIN_TTL, signMessage);
+
+		return signMessage;
+	}
 
-		sendVerificationEmail(user, verifyToken);
+	public async SetCredentialsConfirm (address: string, newPassword: string, signature: string, username: string): Promise<string> {
+		const signMessage = await redisGet(getSetCredentialsKey(address));
+
+		if (!signMessage) {
+			throw new ForbiddenError(messages.SET_CREDENTIALS_SIGN_MESSAGE_EXPIRED);
+		}
+
+		const isValidSr = verifySignature(signMessage, address, signature);
+
+		if (!isValidSr) {
+			throw new ForbiddenError(messages.SET_CREDENTIALS_INVALID_SIGNATURE);
+		}
+
+		const addressObj = await Address
+			.query()
+			.where('address', address)
+			.first();
+
+		if (!addressObj) {
+			throw new ForbiddenError(messages.ADDRESS_NOT_FOUND);
+		}
+
+		const existing = await User
+			.query()
+			.where('username', username.toLowerCase())
+			.first();
+
+		if (existing) {
+			throw new ForbiddenError(messages.USERNAME_ALREADY_EXISTS);
+		}
+
+		const userId = addressObj.user_id;
+		let user = await getUserFromUserId(userId);
+
+		const salt = randomBytes(32);
```

Comment:
This is also duplicated, but not a big deal.

Reply:
removed